# Лабораторная работа 1. Язык Python, основные библиотеки для анализа данных.

ФИО: Николаев Владимир Владимирович

Группа: 317

In [1]:
import numpy as np
import pandas as pd

## Часть 1: Python

#### 1. Напишите код, который сформирует список всех натуральных чисел, не превосходящих 100, которые делятся на 7 и не делятся на 5. Нельзя использовать циклы — выполните это задание с помощью list comprehension.

In [2]:
[x for x in range(101) if not(x % 7) and x % 5]

[7, 14, 21, 28, 42, 49, 56, 63, 77, 84, 91, 98]

#### 2. Напишите функцию `sort_tokens(s)`, которая принимает на вход строку, разделяет ее на токены по запятым, сортирует токены с помощью функции `sorted()` и снова соединяет их в одну строку через запятые. Например, для строки `u"ночь,фонарь,улица"` функция должна выдать `u"ночь,улица,фонарь"`. Придумайте несколько входных строк и протестируйте функцию на них.

Обратите внимание, что склеить строки можно с помощью конструкции `'x'.join(...)`. Например:

In [3]:
'! '.join(['i', 'am', 'very', 'excited', ':-)'])

'i! am! very! excited! :-)'

In [4]:
def sort_tokens(s):
    return ",".join(sorted(s.split(",")))

s1 = u"ночь,фонарь,улица"
print(sort_tokens(s1))

s2 = "aaaa,bbbb,cccc"
print(sort_tokens(s2))

s3 = "aaaa,cccc,bbbb"
print(sort_tokens(s3))

ночь,улица,фонарь
aaaa,bbbb,cccc
aaaa,bbbb,cccc


#### 3. Напишите функцию `remove_duplicates(a)`, которая принимает на вход список и возвращает его же, но без дубликатов и в отсортированном виде. Например, для списка `[u"ночь", u"улица", u"фонарь", u"аптека", u"аптека", u"улица", u"фонарь"]` результат должен быть `[u"аптека", u"ночь", u"улица", u"фонарь", ]`.

Подсказка: обратите внимание на структуру `set`!

In [5]:
def remove_duplicates(a):
    return sorted(set(a))

a = [u"ночь", u"улица", u"фонарь", u"аптека", u"аптека", u"улица", u"фонарь"]
remove_duplicates(a)

['аптека', 'ночь', 'улица', 'фонарь']

#### 4. Напишите функцию `word_counts(texts)`, которая принимает на вход список строк, и печатает пары (слово, число вхождений). Считайте, что слова в текстах могут быть разделены только пробелами. Например, для входа `[u'ночь улица фонарь аптека', u'аптека улица фонарь']` должно быть напечатано
>аптека 2

>улица 2

>фонарь 2

>ночь 1

Подсказки:

Чтобы разбить каждый из текстов на слова, воспользуйтесь функцией `map`, которая применяет заданную функцию к каждому элементу списка. Например, если бы нужно было каждый из текстов привести к нижнему регистру, то команда выглядела бы так:

In [6]:
map(lambda x: x.lower(), ["abCDefg", "HIjkL"])

После первого шага вы получите список, состоящий из списков слов. Чтобы объединить их в один список слов, воспользуйтесь функцией `reduce`. Например, просуммировать числа из списка с ее помощью можно так:



In [7]:
reduce(lambda x, y: x + y, [1, 2, 3, 4, 5])

NameError: name 'reduce' is not defined

Для подсчета количества вхождений слов воспользуйтесь структурой `defaultdict` из модуля `collections`. В отличие от обычного `dict`, она не генерирует ошибку при обращении к несуществущему ключу, а создает новую запись, инициализируемую нулевым значением. Например:

In [8]:
from collections import defaultdict
d = defaultdict(int)
print d
d['abc'] += 1
print d
d['xxx']
print d

SyntaxError: Missing parentheses in call to 'print' (<ipython-input-8-95bf4c99c5b8>, line 3)

In [9]:
from functools import reduce
from collections import Counter

def word_counts(texts):
    #Counter is better than defaultdict for this task imho
    counter = Counter(reduce(lambda x, y: x + y, map(lambda x: x.split(" "), texts)))
    for key in counter:
        print(key, counter[key])

s1 = [u'ночь улица фонарь аптека', u'аптека улица фонарь']
word_counts(s1)

фонарь 2
аптека 2
ночь 1
улица 2


#### 5. (бонусное задание) Скачайте английский перевод "Преступления и наказания". Найдите 20 самых популярных слов и 20 слов, которые встречаются только один раз.

Приведите текст к нижнему регистру. Удалите из него все специальные символы и цифры. Разделите текст на слова, считая разделителями все, кроме букв. Обратите внимание на функцию `split` из модуля `re`.

Не забудьте удалить пустые слова:
```
words = filter(lambda x: len(x) > 0, words)
```

In [10]:
from urllib import urlopen
url = "http://www.gutenberg.org/files/2554/2554.txt"
response = urlopen(url)
raw_text = response.read().decode('utf8')

ImportError: cannot import name 'urlopen'

In [11]:
raw_text[1000:1100]

NameError: name 'raw_text' is not defined

In [12]:
import re
import string
f = open("2554.txt", "r")
raw_text = f.readlines()

raw_text = reduce(lambda x,y: x+y, raw_text, "")
raw_text = raw_text.lower()
allow = string.ascii_lowercase + ' '
raw_text = " " + re.sub('[^%s]' % allow, ' ', raw_text)
words = re.split('[^%s]' % string.ascii_lowercase, raw_text)
words = filter(lambda x: len(x) > 0, words)

print("\nMost common words are:")
counter = Counter(words)
for word, count in counter.most_common(20):
    print(word, count)
    
#There are 1765 words "s" from "something's" i think.

print("\nLeast common words are:")
for word, count in counter.most_common()[:-21:-1]:
    print(word, count)


Most common words are:
the 8003
and 7042
to 5497
he 4900
a 4668
i 4405
you 4118
of 3928
it 3463
that 3308
in 3254
was 2827
his 2116
at 2082
her 1829
but 1801
not 1775
s 1765
with 1757
she 1695

Least common words are:
fronts 1
intrigue 1
accomplished 1
fogeys 1
crippled 1
arrogance 1
flour 1
harbour 1
chord 1
loathsomeness 1
satisfy 1
laziness 1
boozed 1
thrice 1
conductor 1
licensed 1
advisability 1
hem 1
guzzling 1
philosophy 1


1765 слов "s", наверное, от "something's" 

## Часть 2: NumPy и SciPy

#### 1. Повторите 1000 раз следующий эксперимент: сгенерируйте две матрицы размера $10 \times 10$ из стандартного нормального распределения, перемножьте их (как матрицы) и найдите максимальный элемент. Какое среднее значение по экспериментам у максимальных элементов? 95-процентная квантиль?

Для вычисления квантили воспользуйтесь функцией `np.percentile`

In [13]:
avg_max_vector = np.array([])

for it in range(1000):
    m1 = np.random.normal(0, 1, (10, 10))
    m2 = np.random.normal(0, 1, (10, 10))
    avg_max_vector = np.append(avg_max_vector, np.max(np.dot(m1, m2)))
    
print("max: ", np.average(avg_max_vector))
print("95 percentile:", np.percentile(avg_max_vector, 95))

max:  8.38074891238
95 percentile: 11.857989668


#### 2. Сгенерируйте 1000 чисел из распределения Пуассона с параметром $\lambda = 5$. Сформируйте массив, в котором в i-м элементе будет записано количество сгенерированных чисел, равных i.

Обратите внимание на функцию `np.bincount`

In [14]:
a = np.random.poisson(5, 1000)
counter = np.bincount(a)
print(counter)

[  9  44  89 139 180 166 146 103  69  26  18   4   5   1   1]


#### 3. Реализуйте функцию, которая принимает на вход numpy-массив целых чисел `a`, и генерирует массив, в котором число `i` встречается `a[i]` раз. Протестируйте на результате из прошлого задания.

Обратите внимание на функцию `np.repeat`

In [15]:
def repeater(a):
    return np.repeat(range(len(a)), a)

print(repeater(a))

[  0   0   0 ..., 999 999 999]


#### 4. Напишите функцию `scale(X)`, которая принимает на вход матрицу и масштабирует каждый ее столбец (вычитает среднее и делит на стандартное отклонение). Убедитесь, что в функции не будет происходить деления на ноль. Протестируйте на каких-нибудь данных.

In [16]:
def scale(X):
    mean = np.mean(X, axis = 0)
    std = np.std(X, axis = 0)
    std[np.abs(std) < 1e-9] = 1
    X = (X - mean) / std
    return X

x = np.array([[1000, 1], [3000, 1]])
print(scale(x))

[[-1.  0.]
 [ 1.  0.]]


#### 5. Сгенерируйте матрицу с элементами из нормального распределения $\mathcal{N}(10, 1)$. Найдите ее:
- определитель
- след
- наименьший и наибольший элементы
- спектральную норму
- норму Фробениуса
- собственные числа
- обратную матрицу

In [17]:
m = np.random.normal(10, 1, (5, 5))
print("Matrix:\n", m)
print("Det: ", np.linalg.det(m))
print("Trace: ", np.trace(m))
print("Min: ", np.min(m), " Max: ", np.max(m), "")
print("Spectral norm: ", np.linalg.norm(m, ord = 2))
print("Frobenius norm: ", np.linalg.norm(m, ord = "fro"))
print("Eigenvalues:\n", np.linalg.eig(m)[0])
print("Inverse matrix:\n", np.linalg.inv(m))

Matrix:
 [[ 10.03892792   8.9376311   10.32475553   9.81534764  10.16299433]
 [  9.39118525   9.73018197   8.65732608   7.86033742  10.4544594 ]
 [ 10.78062371   9.59434439   9.30813228   9.46645843   7.83514818]
 [ 10.69142665   8.49894575   8.57289031  10.50644708  10.00342088]
 [ 11.22057365   8.72193255   9.6767333    9.50468733  10.99930256]]
Det:  -260.162004638
Trace:  50.5829917975
Min:  7.83514818102  Max:  11.2205736534 
Spectral norm:  48.2329992386
Frobenius norm:  48.3645676328
Eigenvalues:
 [ 48.21090197+0.j          -1.25290395+0.j           0.82790292+1.22548064j
   0.82790292-1.22548064j   1.96918793+0.j        ]
Inverse matrix:
 [[-0.52627438 -0.18253775  0.24408939 -0.17637117  0.64628603]
 [-0.09682485  0.37705774  0.14717693  0.07826803 -0.4449378 ]
 [ 0.48977619 -0.21328356  0.04174791 -0.46659482  0.14479237]
 [ 0.22690547 -0.09310519 -0.10348072  0.50219683 -0.50417507]
 [-0.01331892  0.15531257 -0.31301272  0.09439026  0.09272649]]


## Часть 3: Pandas

#### Ответьте на вопросы о данных по авиарейсам в США за январь-апрель 2008 года.

Данные: http://stat-computing.org/dataexpo/2009/2008.csv.bz2
(обратите внимание, что распаковывать этот файл не обязательно — функция `pandas.read_csv` умеет читать из архивов автоматически)

Описание: http://stat-computing.org/dataexpo/2009/the-data.html

1. Какая из причин отмены рейса (`CancellationCode`) была самой частой? (расшифровки кодов можно найти в описании данных)
2. Найдите среднее, минимальное и максимальное расстояние, пройденное самолетом.
3. Не выглядит ли подозрительным минимальное пройденное расстояние? В какие дни и на каких рейсах оно было? Какое расстояние было пройдено этими же рейсами в другие дни?
4. Из какого аэропорта было произведено больше всего вылетов? В каком городе он находится?
5. Найдите для каждого аэропорта среднее время полета (`AirTime`) по всем вылетевшим из него рейсам. Какой аэропорт имеет наибольшее значение этого показателя?
6. Найдите аэропорт, у которого наибольшая доля задержанных (`DepDelay > 0`) рейсов. Исключите при этом из рассмотрения аэропорты, из которых было отправлено меньше 1000 рейсов (используйте функцию `filter` после `groupby`).

In [18]:
data = pd.read_csv("2008.csv.bz2")

In [19]:
print(data["CancellationCode"].value_counts(dropna=False))
print(len(data))

NaN    6872294
B        54904
A        54330
C        28188
D           12
dtype: int64
7009728


Самой частой причиной были погодные условия (B - weather).

In [20]:
data["Distance"].describe()

count    7009728.000000
mean         726.387029
std          562.101803
min           11.000000
25%          325.000000
50%          581.000000
75%          954.000000
max         4962.000000
Name: Distance, dtype: float64

* Среднее: 726.38 миль
* Минимальное: 11 миль
* Максимальное: 4962 миль

In [21]:
subset = data[(data["Distance"] - 11) < 0.5]
subset[["Year","Month","DayofMonth","DayOfWeek", "FlightNum"]]

Year  Month  DayofMonth  DayOfWeek  FlightNum
2547298  2008      5          15          4       4988
4392215  2008      8          10          7       5572

In [22]:
print("Flight 4988:")
print(data[data["FlightNum"] == 4988]["Distance"].describe())
print("\nFlight 5572:")
print(data[data["FlightNum"] == 4988]["Distance"].describe())

Flight 4988:
count    225.000000
mean     770.208889
std      207.738537
min       11.000000
25%      834.000000
50%      834.000000
75%      903.000000
max      903.000000
Name: Distance, dtype: float64

Flight 5572:
count    225.000000
mean     770.208889
std      207.738537
min       11.000000
25%      834.000000
50%      834.000000
75%      903.000000
max      903.000000
Name: Distance, dtype: float64


Минимальное расстояние в 11 миль было пройдено рейсами 4988 (15/5/2008) и  5572 (10/8/2008).
Среднее растояние на этих рейсах равно 770 милям.

In [23]:
data["Origin"].value_counts()[:1]

ATL    414513
dtype: int64

Больше всего вылетов было произведено из аэропорта ATL ("William B Hartsfield-Atlanta Intl", Atlanta)

In [24]:
mean_air_time = data.groupby("Origin")["Distance"].mean()
print(np.argmax(mean_air_time))

SJU


Наибольшее среднее время полета имеет аэропорт SJU ("Luis Munoz Marin International",San Juan)

In [25]:
data = data.groupby("Origin").filter(lambda x: len(x) >= 1000)

In [26]:
data_delayed = data[data["DepDelay"] > 0].groupby("Origin").size()

In [27]:
data_all = data.groupby("Origin").size()

In [28]:
df = pd.concat([data_all, data_delayed], axis = 1)

In [29]:
np.argmax(df[1] / df[0])

'DAL'

DAL ("Dallas Love", Dallas)- аэропорт с наибольшей долей задержанных рейсов.